Orange/widgets/unsupervised/owpca.py

```diff
@@ -141,6 +141,7 @@ def set_data(self, data):
 
     def fit(self):
         self.clear()
+        self.clear_outputs()
```

Comment:
If data is present, this will first output `None` and then the actual data. Such redundant signals should be avoided because the recepient may have some kind of caching speed-ups.

Method `set_data` already checks whether the table is empty, and clears outputs and exits if so. This would be a proper place to check for `data is None`.

The check in `fit` is needed only for the case when normalization is changed when no data is present.

While you're fixing this: I think that `set_data` should also set `self.data` to `None` in the two cases where it raises errors. This will prevent `fit` from trying to fit when there are no attributes.

Reply:
Thank you for your review, I'll fix this.

Just one thing about this last suggestion regarding the `self.data` setting to `None`. Isn't this already handled in the line 118:
https://github.com/biolab/orange3/blob/2733ef0961c326a2310f732ffb9ae2084d7ee33a/Orange/widgets/unsupervised/owpca.py#L118 
and it doesn't change if on input there is data with no attributes. In this case the if condition `if not data.domain.attributes:` is met and `self.data` will remain `None`

Reply:
> Just one thing about this last suggestion regarding the self.data setting to None. Isn't this already handled in the line 118:

Of course. I'm sorry.